pkg/pgmodel/migrate.go

```diff
@@ -58,99 +69,251 @@ func (t *mySrc) replaceSchemaNames(r io.ReadCloser) (io.ReadCloser, error) {
+		}
+
+		if dbVersion.Compare(version) < 0 && appVersion.Compare(version) >= 0 {
+			if err = execMigrationFiles(tx, filepath.Join("versions", e.Name())); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
 // Migrate performs a database migration to the latest version
 func Migrate(db *sql.DB, versionInfo VersionInfo) (err error) {
 	migrateMutex.Lock()
 	defer migrateMutex.Unlock()
-	// The migration table will be put in the public schema not in any of our schema because we never want to drop it and
-	// our scripts and our last down script drops our shemas
-	driver, err := postgres.WithInstance(db, &postgres.Config{MigrationsTable: "prom_schema_migrations"})
+
+	appVersion, err := semver.Make(versionInfo.Version)
 	if err != nil {
-		return fmt.Errorf("cannot create driver due to %w", err)
+		return fmt.Errorf("app version is not semver format, aborting migration")
 	}
 
-	_, err = db.Exec(timescaleInstall)
+	if err := ensureVersionTable(db); err != nil {
+		return fmt.Errorf("error ensuring version table: %w", err)
+	}
+	dbVersion, dirty, err := getDBVersion(db)
 	if err != nil {
-		return fmt.Errorf("timescaledb failed to install due to %w", err)
+		return fmt.Errorf("failed to get the version from database: %w", err)
 	}
 
-	src, err := httpfs.New(migrations.SqlFiles, "/")
-	if err != nil {
-		return err
+	if dirty {
+		return fmt.Errorf("version found in database dirty, cannot migrate")
+	}
+
+	if err = setDBVersion(db, &appVersion, true); err != nil {
+		return fmt.Errorf("error setting dirty app version to DB: %w", err)
 	}
-	src = &mySrc{src}
 
-	m, err := migrate.NewWithInstance("SqlFiles", src, "Postgresql", driver)
+	tx, err := db.Begin()
 	if err != nil {
-		return err
+		return fmt.Errorf("unable to start transaction: %w", err)
 	}
 	defer func() {
-		sourceErr, databaseErr := m.Close()
-		//don't override error if already set
-		if err != nil {
-			return
-		}
-		if sourceErr != nil {
-			err = sourceErr
-			return
-		}
-		if databaseErr != nil {
-			err = databaseErr
-			return
-		}
+		_ = tx.Rollback()
 	}()
 
-	err = m.Up()
-	//ignore no change errors as we want this idempotent. Being up to date is not a bad thing.
-	if err == migrate.ErrNoChange {
-		err = nil
-	}
+	_, err = tx.Exec(timescaleInstall)
 	if err != nil {
+		return fmt.Errorf("timescaledb failed to install due to %w", err)
+	}
+
+	if dbVersion == nil {
+		if err = execMigrationFiles(tx, preinstallScripts); err != nil {
+			return err
+		}
+	} else {
+		if err = installVersionScripts(tx, dbVersion, &appVersion); err != nil {
+			return err
+		}
+	}
+	if err = execMigrationFiles(tx, idempotentScripts); err != nil {
 		return err
 	}
+	if err = setDBVersion(db, &appVersion, false); err != nil {
+		return fmt.Errorf("error setting clean app version to DB: %w", err)
+	}
+
+	if err = tx.Commit(); err != nil {
+		return fmt.Errorf("unable to commit migration transaction: %w", err)
+	}
 
 	_, extErr := db.Exec(fmt.Sprintf(extensionInstall, extSchema))
```

Comment:
I think some of the issues in the tests are because ExtensionIsInstalled is no longer set